### ISSUE TITLE
Feature: PyTorch-Based Neural Network Learner Plugin with GPU Acceleration

### ISSUE THREAD

**Issue by musaaero** [nasa/OnAIR#198]
## Summary

This issue proposes adding a new **PyTorch Learner plugin** that enables users to run arbitrary PyTorch neural network models inside the OnAIR cognitive pipeline.

Currently, the only implemented Learner plugin is `kalman_plugin.py`, which uses classical Kalman filtering for fault detection. While effective for linear Gaussian systems, it cannot capture the complex, nonlinear relationships present in modern spacecraft telemetry. This proposal fills that gap by introducing a general-purpose deep-learning Learner that plugs directly into the existing `LearnersInterface`.

This complements #150 (TFLite model plugin) — TFLite targets quantized edge/embedded deployment, while this plugin targets the full PyTorch ecosystem for GPU-accelerated inference and research prototyping.

## Motivation

- The `LearnersInterface` (`onair/src/ai_components/learners_interface.py`) is designed to host data-driven AI components, but currently contains only a Kalman filter and a generic placeholder.
- Modern aerospace anomaly detection, state estimation, and fault prediction increasingly rely on deep learning (autoencoders, LSTMs, transformers).
- OnAIR's stated purpose is to "explore research concepts in autonomous operations." Without a neural-network Learner, researchers cannot prototype or evaluate modern AI approaches within the framework.
- PyTorch is the dominant framework in aerospace AI research and supports both CPU inference (flight-like hardware) and GPU acceleration (simulation / ground testing), aligning with OnAIR's goal of scalable simulation.

### New Files

```text
plugins/
  pytorch_learner/
    __init__.py
    pytorch_learner_plugin.py    # Core plugin (inherits AIPlugin)
    model_registry.py            # Utility to load .pt / .onnx model files
onair/config/
  pytorch_learner_example.ini    # Example configuration
test/
  plugins/
    pytorch_learner/
      test_pytorch_learner_plugin.py
doc/
  PytorchLearner.md              # Usage guide and tutorial
```

### Plugin Interface

The plugin will inherit from `AIPlugin` and implement the two required abstract methods:

```python
class Plugin(AIPlugin):
    def __init__(self, name, headers, model_path=None, device="auto",
                 window_size=50, anomaly_threshold=2.0):
        super().__init__(name, headers)
        self.device = self._resolve_device(device)  # "auto" -> cuda if avail, else cpu
        self.model = self._load_model(model_path)
        self.window_size = window_size
        self.anomaly_threshold = anomaly_threshold
        self.buffer = [[] for _ in range(len(headers))]

    def update(self, low_level_data=None, _high_level_data=None):
        'Append new telemetry frame to sliding window buffer.'
        ...

    def render_reasoning(self):
        'Run model inference on buffered window; return anomaly scores and flagged attributes.'
        ...
```

### Key Features

1. **Model-agnostic** — accepts any `torch.nn.Module` saved as `.pt` or exported as `.onnx`. Users supply their own trained model.
2. **GPU/CPU auto-detection** — configurable `device` parameter (`auto`, `cpu`, `cuda`, `cuda:N`). Falls back to CPU gracefully when CUDA is unavailable.
3. **Optional dependency** — PyTorch is **not** added to the base `requirements.txt`. The plugin uses lazy imports and is only functional when `torch` is installed (`pip install onair[pytorch]` or manual install). This keeps OnAIR lightweight for flight-like deployments that do not need neural inference.
4. **Sliding-window inference** — mirrors the Kalman plugin's windowed approach for temporal consistency.
5. **Anomaly scoring** — returns per-attribute anomaly scores via `render_reasoning()`, compatible with the existing `Diagnosis` pipeline in `agent.py`.
6. **Config-driven** — model path, device, window size, and threshold are set via the `.ini` config file, consistent with OnAIR's plugin loading mechanism (`import_plugins` in `plugin_import.py`).
7. **Bounded online updates (optional)** — if online fine-tuning is enabled, gradient updates are clipped and a fixed learning-rate schedule is applied. A `training_mode` flag defaults to `false` to prevent unbounded runtime parameter drift.

### Example Configuration

```ini
[PLUGINS]
LearnersPluginDict = {'pytorch_learner': 'plugins/pytorch_learner'}

[OPTIONS]
IO_Enabled = true
```

Model parameters are passed through the plugin's `__init__` kwargs (consistent with how `kalman_plugin.py` accepts `window_size` and `residual_threshold`).

## Use Cases

- **Autoencoder-based anomaly detection** — train an autoencoder on nominal telemetry; flag frames with high reconstruction error.
- **LSTM/GRU sequence prediction** — predict next telemetry values; flag large residuals (neural alternative to Kalman).
- **Transformer-based fault classification** — classify telemetry windows into fault categories for the `Diagnosis` pipeline.
- **Multi-GPU batch inference** — run multiple model instances across GPUs for high-throughput simulation sweeps.

## Minimal Working Example (MWE)

To demonstrate viability, here is a minimal autoencoder that could be plugged in immediately:

```python
# user_model.py — example model a researcher would train externally
import torch
import torch.nn as nn

class TelemetryAutoencoder(nn.Module):
    def __init__(self, input_dim, latent_dim=8):
        super().__init__()
        self.encoder = nn.Sequential(
            nn.Linear(input_dim, 32), nn.ReLU(),
            nn.Linear(32, latent_dim), nn.ReLU()
        )
        self.decoder = nn.Sequential(
            nn.Linear(latent_dim, 32), nn.ReLU(),
            nn.Linear(32, input_dim)
        )

    def forward(self, x):
        z = self.encoder(x)
        return self.decoder(z)

# Training (external to OnAIR)
# model = TelemetryAutoencoder(input_dim=12)
# ... train on nominal telemetry ...
# torch.save(model.state_dict(), "ae_telemetry.pt")
```

The plugin would load `ae_telemetry.pt`, buffer `window_size` frames, and return reconstruction error as the anomaly score.

## Testing Plan

- **Unit tests** for `update()` buffering and window management (100% line coverage, per repo policy).
- **Unit tests** for `render_reasoning()` with a small dummy `nn.Linear` model.
- **Mock-based tests** for the CUDA path without requiring a GPU (`unittest.mock.patch('torch.cuda.is_available', return_value=True)`).
- **Integration test:** run full `driver.py` simulation with the plugin loaded via `.ini` config against the existing `700_crash_to_earth_1.csv` dataset.
- **Dependency-isolation test:** verify the plugin fails gracefully with an informative error when `torch` is not installed.

## Open Questions for Maintainers

1. Is an extras install (`pip install onair[pytorch]`) acceptable, or would you prefer the plugin live in a separate repo / namespace package?
2. Should `.onnx` loading be included in the initial PR, or deferred to a follow-up?
3. The architecture paper mentions that Learners need "low-level feature data and high-level ground truth labels for training." Should the initial plugin support online updates via `_high_level_data`, or remain inference-only until the ground-truth pipeline is more mature?

## Intention to Work On This

I intend to implement this feature. I am an aeronautical engineering student at Nanjing University of Aeronautics and Astronautics (NUAA) with 3+ years of hands-on PyTorch experience, including multi-GPU training/inference pipelines and deep reinforcement learning (Stable Baselines3, Ray RLlib). I'm really motivated and committed to making this project reach its highest potential.
contact: uma_aero_61@nuaa.edu.cn

## References

- Existing Learner interface: `onair/src/ai_components/learners_interface.py`
- Plugin base class: `onair/src/ai_components/ai_plugin_abstract/ai_plugin.py`
- Kalman plugin (reference implementation): `plugins/kalman/kalman_plugin.py`
- Related issue: #150 (TFLite model plugin)
- OnAIR architecture paper (AIAA SciTech 2024): ["OnAIR: A Cognitive Architecture for Spacecraft Operations"](https://ntrs.nasa.gov/citations/20240008776)
